File: src/client/java/io/github/ricciow/util/TextParser.java
package io.github.ricciow.util;

import net.minecraft.text.*;
import net.minecraft.util.Formatting;

import java.net.URI;

public class TextParser {

    /**
     * Parses a string with legacy '&' color codes into a Minecraft Text component.
     * @param text The string to parse.
     * @return A MutableText object with proper formatting.
     */
    public static MutableText parse(String text) {
        MutableText mainText = Text.literal("");
        String[] parts = text.split("(?=&)"); // Split before each '&' but keep it.
        Style currentStyle = Style.EMPTY;

        for (String part : parts) {
            if (part.startsWith("&") && part.length() > 1) {
                Formatting formatting = Formatting.byCode(part.charAt(1));
                if (formatting != null) {
                    if (formatting.isColor()) {
                        currentStyle = Style.EMPTY.withColor(formatting);
                    } else {
                        currentStyle = currentStyle.withFormatting(formatting);
                    }
                }
                mainText.append(Text.literal(part.substring(2)).setStyle(currentStyle));
            } else {
                mainText.append(Text.literal(part).setStyle(currentStyle));
            }
        }
        return mainText;
    }

    public static MutableText parseHoverable(String text, String hover) {
        MutableText mainText = parse(text);

        HoverEvent hoverEvent = new HoverEvent.ShowText(parse(hover));

        Style hoverStyle = Style.EMPTY.withHoverEvent(hoverEvent);

        mainText.setStyle(hoverStyle);

        return mainText;
    }

    public static MutableText parseHoverable(String text, Text hover) {
        MutableText mainText = parse(text);

        HoverEvent hoverEvent = new HoverEvent.ShowText(hover);

        Style hoverStyle = Style.EMPTY.withHoverEvent(hoverEvent);

        mainText.setStyle(hoverStyle);

        return mainText;
    }
}

File: src/main/kotlin/io/github/ricciow/util/TextParser.kt
package io.github.ricciow.util

import net.minecraft.text.HoverEvent.ShowText
import net.minecraft.text.MutableText
import net.minecraft.text.Style
import net.minecraft.text.Text
import net.minecraft.util.Formatting

object TextParser {
    /**
     * Parses a string with legacy '&' color codes into a Minecraft Text component.
     * @param text The string to parse.
     * @return A MutableText object with proper formatting.
     */
    fun parse(text: String): MutableText {
        val mainText = "".toText()
        val parts = text.split("(?=[&§])".toRegex()) // Split before each '&' but keep it.
        var currentStyle = Style.EMPTY

        for (part in parts) {
            if ((part.startsWith("&") || part.startsWith("§")) && part.length > 1) {
                val formatting = Formatting.byCode(part[1])
                if (formatting != null) {
                    currentStyle = if (formatting.isColor) {
                        Style.EMPTY.withColor(formatting)
                    } else {
                        currentStyle.withFormatting(formatting)
                    }
                }
                mainText.append(part.substring(2).toText(currentStyle))
            } else {
                mainText.append(part.toText(currentStyle))
            }
        }
        return mainText
    }

    fun parseHoverable(text: String, hover: String): MutableText {
        return parse(text).apply {
            style = Style.EMPTY.withHoverEvent(ShowText(parse(hover)))
        }
    }

    fun parseHoverable(text: String, hover: Text): MutableText {
        return parse(text).apply {
            style = Style.EMPTY.withHoverEvent(ShowText(hover))
        }
    }
}